meadow-client/src/MonadApp.purs

```diff
@@ -0,0 +1,394 @@
+    updateContractInState
+      ( case newContract of
+        Nothing -> ""
+        Just x -> x
+      )
+  saveBuffer text = wrap $ liftEffect $ LocalStorage.setItem bufferLocalStorageKey text
+  saveMarloweBuffer text = wrap $ liftEffect $ LocalStorage.setItem marloweBufferLocalStorageKey text
+  getOauthStatus = runAjax Server.getOauthStatus
+  getGistByGistId gistId = runAjax $ Server.getGistsByGistId gistId
+  postGist newGist = runAjax $ Server.postGists newGist
+  patchGistByGistId newGist gistId = runAjax $ Server.patchGistsByGistId newGist gistId
+  postContractHaskell source = runAjax $ Server.postContractHaskell source
+
+runHalogenApp :: forall m a. HalogenApp m a -> HalogenM FrontendState Query (Coproduct AceQuery AceQuery) (Either EditorSlot MarloweEditorSlot) Void m a
+runHalogenApp = unwrap
+
+runAjax ::
+  forall m a.
+  ExceptT AjaxError (HalogenM FrontendState Query (Coproduct AceQuery AceQuery) (Either EditorSlot MarloweEditorSlot) Void m) a ->
+  HalogenApp m (WebData a)
+runAjax action = wrap $ RemoteData.fromEither <$> runExceptT action
+
+-- | Handles the messy business of running an editor command if the
+-- editor is up and running.
+withEditor ::
+  forall m a.
+  MonadEffect m =>
+  (Editor -> Effect a) ->
+  HalogenApp m (Maybe a)
+withEditor action =
+  HalogenApp
+    $ do
+        mEditor <- query' cpEditor EditorSlot $ request GetEditor
+        case join mEditor of
+          Just editor -> Just <$> (liftEffect $ action editor)
+          _ -> pure Nothing
+
+withMarloweEditor ::
+  forall m a.
+  MonadEffect m =>
+  (Editor -> Effect a) ->
+  HalogenApp m (Maybe a)
+withMarloweEditor action =
+  HalogenApp
+    $ do
+        mEditor <- query' cpMarloweEditor MarloweEditorSlot $ request GetEditor
+        case mEditor of
+          Just (Just editor) -> do
+            liftEffect $ Just <$> action editor
+          _ -> pure Nothing
+
+updateContractInStateP :: String -> MarloweState -> MarloweState
+updateContractInStateP text state = set _contract con state
+  where
+  con = case runParser text contract of
+    Right pcon -> Just pcon
+    Left _ -> Nothing
+
+updateStateP :: MarloweState -> MarloweState
+updateStateP oldState = actState
+  where
+  sigState = updateSignatures oldState
+
+  mSimulatedState = simulateState sigState
+
+  actState = case mSimulatedState of
+    Just simulatedState -> updateActions sigState simulatedState
+    Nothing -> sigState
+
+updateSignatures :: MarloweState -> MarloweState
+updateSignatures oldState = case oldState.contract of
+  Just oldContract -> over (_transaction <<< _signatures) (resizeSigs (peopleFromStateAndContract (oldState.state) oldContract)) oldState
+  Nothing -> oldState
+
+simulateState :: MarloweState -> Maybe {state :: State, contract :: Contract, outcome :: TransactionOutcomes, validity :: TransactionValidity}
+simulateState state = case state.contract of
+  Just c ->
+    Just
+      ( case applyTransaction inps sigs bn st c mic of
```

Comment:
I'd be very tempted to roll the if into the pattern match:

```
case inps, applyTransaction inps sigs bn st c mic of
  _, MCouldNotApply err -> ...
  Nil, MCouldNotApply InvalidInput -> ...
  _, _ -> ...
```

Might be one for a separate PR though. 🤔

Reply:
done